### ISSUE TITLE
Implementing NIST 800-63b policy

### ISSUE THREAD

**Issue by justinschoeman** [libpwquality/libpwquality#63]
This probably does not qualify as an issue - more of an enhancement request.

I would like to implement the NIST 800-63b policy for memorized secrets.  The tricky bit is:

"Verifiers SHOULD NOT require memorized secrets to be changed arbitrarily (e.g., periodically). However, verifiers SHALL force a change if there is evidence of compromise of the authenticator."

I have a script which grabs a number of breach databases and builds an updated cracklib dictionary. So I can test passwords to see if they appear in any of these databases (evidence of a compromise).

The issue is that pwquality only runs when setting a new password, and the NIST standard explicitly recommends against regular password changes. So I would need to find a way to run pwquality checks in the auth chain in order to regularly check them against the breach databases.

Does such a module exist? Or can pam_pwquality be extended to provide such functionality?

Thanks,
Justin

**Comment by t8m** [comment]
> Does such a module exist? Or can pam_pwquality be extended to provide such functionality?

I do not know about such module. Of course pam_pwquality could be extended to be applied also in the pam_authenticate() step. The question is what would be the outcome. Maybe the pam_authenticate() step could be combined with pam_acct_mgmt() - i.e., to return PAM_NEW_AUTHTOK_REQD there to force the user to change their password.

**Comment by justinschoeman** [comment]
That is where it gets interesting... It would probably have to go after pam_unix (you want to check the actual successful password) - but that is already 'sufficient'.  Would any return code after that be sufficient to trigger a new password request?

(It has been 15 years since I wrote a PAM module, and things were a lot simpler than ;) .)

From my side, the most important thing would be a syslog entry, which gets picked up by logwatch and alerted (a freshly compromised password should warrant an investigation).  It would be nice if the user could be forced to change their password too - but at least for my use case, that is not essential.

**Comment by t8m** [comment]
Yeah pam_unix couldn't be sufficient otherwise modules after it would be skipped.
